### ISSUE TITLE
Ability to fine-tune socket options (timeout, keepalive, linger, etc.)

### ISSUE THREAD

**Issue by oranoran** [hierynomus/smbj#91]
It would be useful to be able to set socket options for the sockets used by SMB connections.
Specifically, SO_TIMEOUT, and probably KEEPALIVE and SO_LINGER seem to be useful.

I created a [potential fix](https://github.com/intigua/smbj/tree/socket-options), allowing Config to receive a customized SocketFactory, and allowing ProxySocketFactory to be customized with some socket options.

The only problem is that when I'm testing this fix with the following code, even though sockets get set with the desired socket options, their behavior doesn't change at all. Any ideas?

```
@Test
public void checkFactorySocketTimeout() throws InterruptedException, IOException {
    ProxySocketFactory factory = new ProxySocketFactory();
    factory.setSoTimeout(100);
    factory.setKeepAlive(true);
    Config config = Config.builder().withSocketFactory(factory).build();
    Connection connection = new Connection(config, new SMBEventBus());
    connection.connect(myhost);
    Thread.sleep(200);
    assertFalse(connection.isConnected());
}
```

**Comment by hierynomus** [comment]
There is no need to inject/change `SocketFactory`. The `SocketClient` actually takes care of setting `soTimeout`, and should take care of the rest also. Adding those as config parameters makes sense indeed, I was thinking the same when Pepijn sent the PR this weekend.

**Comment by oranoran** [comment]
Good catch, somehow I missed it.
And unlike my change, this works, though I'm not sure what the difference is.

But adding connection.setSotTimeout(100) to my test still fails the test, since connection.isConnected() continues to return true. The timeout event kills the PacketReader thread, but it doesn't cause the socket to be closed Wouldn't it make sense to close the socket and set it to null in this case, since there is now no thread that can read it?

**Comment by hierynomus** [comment]
Yes, I think that makes sense.

**Comment by oranoran** [comment]
Ok, I will try to create a PR for this (hopefully tomorrow). Probably should also verify that an SO timeout causes Future.get() requests to fail.

**Comment by hierynomus** [comment]
It wouldn't directly fail the Futures, as only when a Packet is received we know for which Future it is. It will however close complete all the waiting futures if we close the connection, due to the SO Timeout.

**Comment by oranoran** [comment]
About using setSoTimeout - I'm not sure whether calling this method will have an effect if called after the socket is already connected. In case it doesn't have an effect, then the only way to effectively set the timeout is not to use SMBClient and create connections directly.

A simpler change then would be for soTimeout to be part of the config, and then SMBClient.connect() method should set it. Makes sense?

**Comment by hierynomus** [comment]
You can call it after a connect, but it should be called before a blocking operation according to the javadoc.

This is added in fee525c

I'm doubting whether any of the other options should be exposed. What do you think? I would be in favour of not exposing them unless absolutely necessary.

**Comment by oranoran** [comment]
I think keepalive is probably the next-most-useful one, in case someone wants to leave the client running idle for a while and ensure liveness. But it's less important then soTimeout.

There are quite a few other TCP connection parameters, though as you say most of them seem not interesting. One way of keeping TCP configuration flexible is #93 - exposing the SocketFactory in the Config. This way any user can set up a custom socket factory if they need, controlling any parameter that they want.
But I realize the price of this in terms of loss of abstraction, SocketFactory is a more low-level notion. I don't know if there is a more appropriate place than Config for customizing the SocketFactory.

Bottom line, I  don't see a strong need for any further extensions right now.
